client/src/context/AuthContext.tsx

```diff
@@ -11,76 +11,160 @@ import {
 } from 'firebase/auth';
 import Cookies from 'js-cookie';
 
+const { auth } = app;
 const csrftoken = Cookies.get('csrftoken');
 
 export interface User {
   username: string, 
   email: string, 
-  uuid: string | unknown,
-  location: number | unknown
+  uuid: string | null,
+  location: string
 };
 
 interface AuthenticatedUser {
-  signup: (username: string, email: string, password: string) => Promise<UserCredential>;
+  signup: (username: string, email: string, password: string) => Promise<UserCredential | undefined>;
   login: (email: string, password: string) => Promise<UserCredential>;
   logout: () => Promise<void>;
   resetPassword: (email: string) => Promise<void>;
   user: User;
+  refreshUser: (userId: string, email: string | null) => Promise<void>;
+  dispatchUser: React.Dispatch<UserReducerAction>;
   csrftoken: string | undefined;
+};
+
+interface UserReducerAction {
+  type: 'set_user' | 'clear_user' | 'set_username' | 'set_email' | 'set_uuid' | 'set_location',
+  newUsername?: string,
+  newEmail?: string,
+  newUuid?: string,
+  newLocation?: string,
+  newUserState?: User
+}
+
+function userReducer(state: User, action: UserReducerAction): User {
+  switch (action.type) {
+    case 'set_user': {
+      return {
+        ...state,
+        username: action.newUsername ?? state.username,
+        email: action.newEmail ?? state.email,
+        uuid: action.newUuid ?? state.uuid,
+        location: action.newLocation ?? state.location
+      }
+    }
+    case 'clear_user': {
+      return {
+        username: '',
+        email: '',
+        uuid: null,
+        location: '13'
+      };
+    }
+    case 'set_username': {
+      return {
+        ...state,
+        username: action.newUsername ?? state.username
+      };
+    }
+    case 'set_email': {
+      return {
+        ...state,
+        email: action.newEmail ?? state.email
+      };
+    }
+    case 'set_uuid': {
+      return {
+        ...state,
+        uuid: action.newUuid ?? state.uuid
+      };
+    }
+    case 'set_location': {
+      return {
+        ...state,
+        location: action.newLocation ?? state.location
+      };
+    }
+  }
 }
 
 const UserContext = createContext<AuthenticatedUser | null>(null);
 
-  export const AuthContextProvider = ({ children }: { children: ReactNode }) => {
-    const [user, setUser] = useState<User>({
+export const AuthContextProvider = ({ children }: { children: ReactNode }) => {
+  const [user, dispatchUser] = useReducer(userReducer,
+    {
       username: '',
       email: '',
       uuid: null,
-      location: 1
-    });
+      location: "13"
+    }
+  );
```

Comment:
`useReducer` is basically a fancier version of `useState`
[React Docs](https://react.dev/reference/react/useReducer#reference)